File: hotel-mern/services/type.service.ts
import { TypeRepository } from '../shared/repositories'
import { Type } from '../shared/models'

export default class TypeService extends TypeRepository {
  async get(): Promise<Type[]> {
    return this.model.find({}).lean()
  }

  async create(
    _services: string[],
    name: string,
    places: number
  ): Promise<Type> {
    const type = await new this.model({
      _services,
      name,
      places,
    })
    await type.save()
    return type
  }

  async change(
    _id: string,
    _services: string[],
    name: string,
    places: number
  ): Promise<Type> {
    await this.model
      .findByIdAndUpdate(_id, {
        $set: { _services, name, places },
      })
      .lean()
    return this.model.findById(_id).lean()
  }

  async removeService(serviceId: string) {
    await this.model.updateMany({}, { $pull: { _services: serviceId } })
  }

  async delete(_id: string): Promise<string> {
    await this.model.findByIdAndRemove(_id)
    return _id
  }
}

File: hotel/backend/src/main/kotlin/loshica/api/hotel/services/TypeService.kt
package loshica.api.hotel.services

import loshica.api.hotel.core.BaseService
import loshica.api.hotel.interfaces.ITypeService
import loshica.api.hotel.models.Option
import loshica.api.hotel.models.Room
import loshica.api.hotel.models.Type
import loshica.api.hotel.repositories.TypeRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.util.*

@Service
class TypeService(
    @Autowired override val repository: TypeRepository
) : BaseService<Type, TypeRepository>(repository), ITypeService {

    override fun create(
        name: String,
        places: Int,
        price: Int,
        options: List<Option>
    ): Type {
        val type = Type(
            name = name,
            places = places,
            price = price,
            options = options.toMutableList()
        )
        repository.save(type)
        return type
    }

    override fun change(
        id: Int,
        name: String,
        places: Int,
        price: Int,
        options: List<Option>
    ): Type {
        val type: Type = getOne(id)
        type.change(name = name, places = places, price = price, options = options)
        repository.save(type)
        return type
    }

    override fun removeOption(option: Option) = getAll().forEach {
        it.options.remove(option)
        repository.save(it)
    }

    override fun addRoom(room: Room) {
        room.type.let {
            it.rooms.add(room)
            repository.save(it)
        }
    }

    override fun removeRoom(room: Room) {
        room.type.let {
            it.rooms.remove(room)
            repository.save(it)
        }
    }

    override fun clearOptions(type: Type) {
        type.options = mutableListOf()
        repository.save(type)
    }

    override fun disable(id: Int): Int {
        val type: Type = getOne(id)
        type.isActive = false
        type.name = "${Date().time}-${Math.random()}"
        repository.save(type)
        return id
    }
}